Detectors/EMCAL/simulation/src/Detector.cxx

```diff
@@ -135,22 +137,36 @@ Bool_t Detector::ProcessHits(FairVolume* v)
   LOG(DEBUG3) << "Name of the supermodule type " << mcapp->CurrentVolOffName(4) << ", Module name "
               << mcapp->CurrentVolOffName(3) << std::endl;
 
-  TLorentzVector pos, mom;
-  mcapp->TrackPosition(pos);
-  mcapp->TrackMomentum(mom);
-  Double_t estart = mcapp->Etot(), time = mcapp->TrackTime() * 1e9; // time in ns
   Int_t partID = mcapp->GetStack()->GetCurrentTrackNumber(),
-        parent = mcapp->GetStack()->GetCurrentTrack()->GetMother(0),
-        detID = geom->GetAbsCellId(offset + copySmod - 1, copyMod - 1, copyPhi - 1, copyEta - 1);
+    parent = mcapp->GetStack()->GetCurrentTrack()->GetMother(0),
+    detID = geom->GetAbsCellId(offset + copySmod - 1, copyMod - 1, copyPhi - 1, copyEta - 1);
 
   Double_t lightyield(eloss);
   if (mcapp->TrackCharge())
     lightyield = CalculateLightYield(eloss, mcapp->TrackStep(), mcapp->TrackCharge());
   lightyield /= geom->GetSampling();
 
-  /// check handling of primary particles
-  AddHit(partID, parent, 0, estart, detID, Point3D<float>(pos.X(), pos.Y(), pos.Z()),
-         Vector3D<float>(mom.Px(), mom.Py(), mom.Pz()), time, lightyield);
+  if(partID != mCurrentTrackID || detID != mCurrentCellID || !mCurrentHit){
+    // Condition for new hit:
+    // - Processing different track
+    // - Inside different cell
+    // - First track of the event
+    std::cout << "New track / cell started\n";
+    TLorentzVector pos, mom;
```

Comment:
a tiny optimization hint: This will create (construct) TLorentzVector again and again. (I have seen this is the past). Consider making mPosCache, mMomCache private member variables of the detector.

Reply:
Even less important (especially if you then use it as a data member of the class): consider using [ROOT::Math::LorentzVector](https://root.cern.ch/doc/master/classROOT_1_1Math_1_1LorentzVector.html), most of the time it can be used as a drop-in replacement of TLorentzVector, it does not have the TObject overhead and you can use the float version for half the size in memory.

Reply:
@mpuccio : How can this be used if the VMC interface expects TLorentzVector?

Reply:
OK, could be done, although I normally prefer to not have local variables as local variables. On the long term, in this case maybe one could discuss with Ivana to add the same member functions with primitive output types (array of doubles) in order to stay minimalistic. All we need is the x and p vector, the functionality of TLorentzVector stays largely unused.